source/Timer.cpp

```diff
@@ -0,0 +1,326 @@
+	result.resetCondition = resetCondition;
+	result.repeatReset = repeatReset;
+	result.resetFired = resetFired;
+	result.idleMaxSpeed = idleMaxSpeed;
+
+	result.timeElapsed = timeElapsed;
+	result.isComplete = isComplete;
+	result.isActive = isActive;
+
+	for(const auto &it : actions)
+		result.actions[it.first] = it.second.Instantiate(subs, origin, jumps, payload);
+
+	if(rand >= 1)
+		result.timeToWait = timeToWait + Random::Int(rand);
+	else
+		result.timeToWait = timeToWait;
+
+	return result;
+}
+
+
+
+bool Timer::IsComplete() const
+{
+	return isComplete;
+}
+
+
+
+bool Timer::IsOptional() const
+{
+	return optional;
+}
+
+
+
+void Timer::ResetOn(ResetCondition cond, PlayerInfo &player, UI *ui, Mission *mission)
+{
+	bool reset = cond == resetCondition;
+	reset |= (cond == Timer::ResetCondition::LEAVE_ZONE && resetCondition == Timer::ResetCondition::PAUSE);
+	reset |= (cond == Timer::ResetCondition::LEAVE_SYSTEM && (resetCondition == Timer::ResetCondition::PAUSE
+				|| resetCondition == Timer::ResetCondition::LEAVE_ZONE));
+	if(isActive && reset)
+	{
+		timeToWait = timeToWait + timeElapsed;
+		timeElapsed = 0;
+		if(repeatReset || !resetFired)
+		{
+			actions[RESET].Do(player, ui, mission);
+			resetFired = true;
+		}
+		isActive = false;
+	}
+}
+
+
+
+void Timer::Step(PlayerInfo &player, UI *ui, Mission *mission)
+{
+	if(isComplete)
+		return;
+	Ship *flagship = player.Flagship();
+	if(!flagship)
+		return;
+	if((system && flagship->GetSystem() != system) ||
+		(!systems.IsEmpty() && !systems.Matches(flagship->GetSystem())))
+	{
+		ResetOn(Timer::ResetCondition::LEAVE_SYSTEM, player, ui, mission);
+		return;
+	}
+	if(requireIdle)
+	{
+		bool shipIdle = (!flagship->IsThrusting() && !flagship->IsSteering()
+						&& !flagship->IsReversing() && flagship->Velocity().LengthSquared() < idleMaxSpeed);
+		for(const Hardpoint &weapon : flagship->Weapons())
+			shipIdle &= !weapon.WasFiring();
+		if(!shipIdle)
+		{
+			ResetOn(Timer::ResetCondition::PAUSE, player, ui, mission);
+			return;
+		}
+	}
+	if(requireUncloaked)
+	{
+		double cloak = flagship->Cloaking();
+		if(cloak != 0.)
+		{
+			ResetOn(Timer::ResetCondition::PAUSE, player, ui, mission);
+			return;
+		}
+	}
+	if(proximity > 0.)
+	{
+		bool inProximity = false;
+		if(proximityCenter || !proximityCenters.IsEmpty())
+		{
+			for(const StellarObject &proximityObject : system->Objects())
+				if(proximityObject.HasValidPlanet() && (proximityCenter == proximityObject.GetPlanet() ||
+					proximityCenters.Matches(proximityObject.GetPlanet())))
+				{
+					double dist = flagship->Position().Distance(proximityObject.Position());
+					if((closeTo && dist <= proximity) || (!closeTo && dist >= proximity))
+					{
+						inProximity = true;
+						break;
+					}
+				}
+		}
+		if(!inProximity)
+		{
+			ResetOn(Timer::ResetCondition::LEAVE_ZONE, player, ui, mission);
+			return;
+		}
+	}
+	isActive = true;
+	if(++timeElapsed >= timeToWait)
+	{
+		actions[TIMEUP].Do(player, ui, mission);
```

Comment:
There may not be a TIMEUP action defined.
```suggestion
		auto it = actions.find(TIMEUP)
		if(it != actions.end())
			it->second.Do(player, ui, &mission);
```